### PR TITLE
Clementine-playlists to Library integration

### PR DESCRIPTION
**Pull request by julled**
Integration of playlist usage from the clementine player (the one and only good player for linux ;)). I was using the banshee interface as reference and adapted it to the sqlite clementine lib. As they share some similarities it would be possible to refactor it to a more generalized sqlite interface, but for now its also okey this way for me.
Implemented / tested in ubuntu 16.04, but it should work under windows/macos aswell (untested but implemented).
A logo in the same style as the others needs to be added to resources. 

i also just signed the Mixxx Contributor Agreement.

(Its my first pullrequest to a opensource project ever, so if i missed something, please give me a hint :))

### CODE REVIEW

**Comment by julled** [comment]
Okey i am working on that, will give you an update tomorrow.

**Comment by daschuer** [comment]
Thank you. You van you keep this branch and OR open, if one likes to use it on top of current Mixxx.

File: src/library/clementine/clementinedbconnection.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            entry.grouping = query.value(15).toString();
+            entry.tracknumber = query.value(9).toInt();
+            entry.bpm = query.value(10).toInt();
+            entry.bitrate = query.value(11).toInt();
+            entry.comment = query.value(12).toString();
+            entry.playcount = query.value(13).toInt();
+            entry.composer = query.value(14).toString();
+
+            if(entry.artist == "" && entry.title == ""){
+            	entry.artist = "Unknown";
```
Comment (daschuer) [inline src/library/clementine/clementinedbconnection.cpp]: Im afraid that "Unknown" will be finally witten to the file tags if we replace it here.
Why do we need it?

File: src/library/clementine/clementinedbconnection.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            	entry.title = entry.uri.toString().split(QDir::separator()).last();
+            }
+            if(entry.bpm == -1){
```
Comment (daschuer) [inline src/library/clementine/clementinedbconnection.cpp]: I assume -1 means unknown while 0 means beat free. What is the reason for this change?

File: src/library/clementine/clementinedbconnection.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            	entry.bpm=0;
+            }
+            if(entry.duration == -1){
```
Comment (daschuer) [inline src/library/clementine/clementinedbconnection.cpp]: The same here -1 unknown. 0 = empty track.

**Comment by Be-ing** [comment]
> Thank you. You van you keep this branch and OR open, if one likes to use it on top of current Mixxx.

Yeah you may as well keep this branch around for now. Start a new branch from this before rebasing. If you need help with Git feel free to ask on [Zulip](https://mixxx.zulipchat.com/).

File: src/library/clementine/clementinedbconnection.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    m_database.setDatabaseName(databaseFile);
+    m_database.setConnectOptions("SQLITE_OPEN_READONLY");
+
+    //Open the database connection in this thread.
+    if (!m_database.open()) {
+        m_database.setConnectOptions(); // clear options
+        qWarning() << "Failed to open Clementine database." << m_database.lastError();
+        return false;
+    } else {
+        // TODO(DSC): Verify schema
```
[the remark this answers is not in the document]
Comment (julled) [reply src/library/clementine/clementinedbconnection.cpp]: i basically copied the code from the banshee feature on the master branch where this comment was placed. But i think its not valid anymore for this code. Will remove.

File: src/library/clementine/clementinedbconnection.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    m_database.close();
+}
+
+bool ClementineDbConnection::open(const QString& databaseFile) {
+    m_database = QSqlDatabase::addDatabase("QSQLITE");//, "BANSHE_DB_CONNECTION");
+    m_database.setHostName("localhost");
+    m_database.setDatabaseName(databaseFile);
+    m_database.setConnectOptions("SQLITE_OPEN_READONLY");
+
+    //Open the database connection in this thread.
```
[the remark this answers is not in the document]
Comment (julled) [reply src/library/clementine/clementinedbconnection.cpp]: yes its my issue.

File: src/library/clementine/clementinedbconnection.cpp (hunk 1 above)
Comment (julled) [reply src/library/clementine/clementinedbconnection.cpp]: There are some entrys in the clementine database which get corrupted in a way that most of the informations about the track is missing, even if the filedirectory is still valid and the entrys work in Clementine. I dont know the reason for that, but i was choosing the way to just use the filename as tracktitle and "unknown" as Artist, for better readability and completeness of the playlist, even if the track is not valid anymore in the clementine playlist.
As far as i understood, the import of the playlists are just temporary in the mixxx-internal database for each startup and no information is written back to the id tags. Am i right?

File: src/library/clementine/clementinedbconnection.cpp (hunk 2 above)
Comment (julled) [reply src/library/clementine/clementinedbconnection.cpp]: In the clementine database there is either -1 (unknown) or the bpm. So i chose to change the bpm to zero as all the not analysed tracks in mixxx also got zero.

File: src/library/clementine/clementinedbconnection.cpp (hunk 3 above)
Comment (julled) [reply src/library/clementine/clementinedbconnection.cpp]: As written above there are some corrupt entrys in the database, so i chose to just write no duration but keep the entry for the completeness of the list.

**Comment by julled** [comment]
Hurray, i destroyed my local branch while trying to rebase, but i could get it from my github account back... i think i will just keep this patch for people who want to use the integration on the master.

But for a more sustainable integration i will create a new pull request of my changes on the jmigual-library-redesign branch and fill up a new pull request. Is that okey?

**Comment by daschuer** [comment]
Yes, sounds good.

**Comment by daschuer** [comment]
This would work for me as well, the work is almost the same. 
Merge this to master, adopt the changes to #1726 when merging master into it.

**Comment by Be-ing** [comment]
@daschuer are you volunteering to fix the merge conflicts then? I expect this to require substantial rewriting to merge with the library redesign branch.

**Comment by daschuer** [comment]
No problem. A second branch on top of the library redesign will be very helpful though.
@julled : are you working in it right now?

**Comment by julled** [comment]
Yes i was porting it already to the lib redesign branch, it wasnt lots of work, just some work on the constructors and on some inherited member functions. 
So should i push my further fixes to the master branch or should i open a new pull request for the feature to the redisgn branch https://github.com/mixxxdj/mixxx/tree/jmigual-library-redesign ?

**Comment by Be-ing** [comment]
Please open a new pull request for your new branch. Make sure the new pull request is targeted at the jmigual-library-redesign branch of the mixxxdj/mixxx repository.

**Comment by julled** [comment]
Made that patch for the jmigual-library-redesign branch in #1764.

**Comment by julled** [comment]
continued work on this in #3173
